### PR TITLE
Introduce C and C++ APIs for Sparse Tensors

### PR DESCRIPTION
**Pull request by yuslepukhin**
Add IsSparseTensor
Add CreateSparseTensor*
Add utilities and test fully sparse instantiation
Fully sparse blocksparse
Add test and docs for fully sparse tensor instantiation
Rework creation API
Use API
Non string API
Retrofit of existing String API
Add tests
Add documentation

**Description**: Describe your changes.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

**Comment by yuslepukhin** [comment]
> Status SparseTensor::MakeCooData(const IDataTransfer& data_transfer,

Prohibit strings in Make()

---
Refers to: onnxruntime/core/framework/sparse_tensor.cc:277 in e718ed5. [](commit_id = e718ed5c4b47f2c647ab04e95c56eb76c979bb70, deletion_comment = False)

**Comment by yuslepukhin** [comment]
>  * Get the type information of an OrtValue

Document for Sparse

---
Refers to: include/onnxruntime/core/session/onnxruntime_c_api.h:681 in e718ed5. [](commit_id = e718ed5c4b47f2c647ab04e95c56eb76c979bb70, deletion_comment = False)

**Comment by yuslepukhin** [comment]
>      * \param s string contents. Each string is NOT null-terminated.

Document for sparse

---
Refers to: include/onnxruntime/core/session/onnxruntime_c_api.h:622 in e718ed5. [](commit_id = e718ed5c4b47f2c647ab04e95c56eb76c979bb70, deletion_comment = False)

File: onnxruntime/core/session/ort_apis.h [hunk 1]
```diff
@@ -288,4 +288,31 @@ ORT_API(void, ReleaseTensorRTProviderOptions, _Frees_ptr_opt_ OrtTensorRTProvide
 ORT_API_STATUS_IMPL(EnableOrtCustomOps, _Inout_ OrtSessionOptions* options);
 ORT_API_STATUS_IMPL(RegisterAllocator, _Inout_ OrtEnv* env, _In_ OrtAllocator* allocator);
 ORT_API_STATUS_IMPL(UnregisterAllocator, _Inout_ OrtEnv* env, _In_ const OrtMemoryInfo* mem_info);
+// SparseTensor related API
+ORT_API_STATUS_IMPL(IsSparseTensor, _In_ const OrtValue* value, _Out_ int* out);
+ORT_API_STATUS_IMPL(CreateSparseTensorAsOrtValue, _Inout_ OrtAllocator* allocator, _In_ const int64_t* dense_shape,
+                    size_t dense_shape_len, ONNXTensorElementDataType type, _Outptr_ OrtValue** out);
+ORT_API_STATUS_IMPL(FillSparseTensorCoo, _Inout_ OrtValue* ort_value, _In_ const OrtMemoryInfo* mem_info,
+                    _In_ const int64_t* values_shape, size_t values_shape_len, _In_ const void* values,
+                    size_t indices_num, const int64_t* indices_data);
```
Comment (snnn) [inline onnxruntime/core/session/ort_apis.h:297]: nit: you may add "_In_ " in front of "const int64_t* indices_data.

File: onnxruntime/core/session/ort_apis.h [hunk 2]
```diff
@@ -288,4 +288,31 @@ ORT_API(void, ReleaseTensorRTProviderOptions, _Frees_ptr_opt_ OrtTensorRTProvide
...
+ORT_API_STATUS_IMPL(FillSparseTensorCsr, _Inout_ OrtValue* ort_value, _In_ const OrtMemoryInfo* data_mem_info,
+                    const int64_t* values_shape, size_t values_shape_len, const void* values,
+                    size_t inner_indices_num, const int64_t* inner_indices_data,
+                    size_t outer_indices_num, const int64_t* outer_indices_data);
```
Comment (snnn) [inline onnxruntime/core/session/ort_apis.h:301]: Save as the above.  You may add "_In_ " to the const pointers . Such annotations can help on static code analysis. Not required, just a recommendation.

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 3]
```diff
@@ -589,14 +594,22 @@ struct OrtApi {
...
   /**
+     * Obtain a total length of strings contained within a tensor.
+     * For sparse tensors it returns the total length of values (nnz) strings.
      * \param value A tensor created from OrtCreateTensor... function.
      * \param len total data length, not including the trailing '\0' chars.
      */
   ORT_API2_STATUS(GetStringTensorDataLength, _In_ const OrtValue* value, _Out_ size_t* len);
 
   /**
+     * This API returns a all of of UTF-8 encoded strings that are contained within a tensor
```
Comment (yuslepukhin) [inline include/onnxruntime/core/session/onnxruntime_c_api.h:605]: >a [](http://example.com/codeflow?start=24&length=1)

del

Comment (snnn) [inline include/onnxruntime/core/session/onnxruntime_c_api.h:605]: Thank for you adding this.

**Review by snnn — DISMISSED** [review]

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 4]
```diff
@@ -1489,6 +1512,205 @@ struct OrtApi {
...
+   *  at the creation time has memory info that is not the same as mem_info argument to this function a X-device copy will be performed.
+   *  String data is assumed to be on CPU and will only be copied into a CPU allocated buffer.
+   * \param values - structure with values information
+   * \param indices_shape_len - length of the block sparse indices shape
+   * \param indices_shape - pointer to a location of indices shape
+   * \param indices_data - pointer to a location of indices data. Shape will determine the length of the indices data.
+   */
+  ORT_API2_STATUS(FillSparseTensorBlockSparse, _Inout_ OrtValue* ort_value, _In_ const OrtMemoryInfo* data_mem_info,
+                  _In_ const int64_t* values_shape, size_t values_shape_len, _In_ const void* values,
+                  size_t indices_shape_len, const int64_t* indices_shape_data,
```
Comment (pranavsharma) [inline include/onnxruntime/core/session/onnxruntime_c_api.h:1596]: Would it be better for our users if we remain consistent with order of ptr and len?

**Review by pranavsharma — APPROVED** [review]

**Review by pranavsharma — DISMISSED** [review]

File: include/onnxruntime/core/session/onnxruntime_c_api.h (hunk 4 above)
Comment (pranavsharma) [reply include/onnxruntime/core/session/onnxruntime_c_api.h]: Thx for addressing this.

**Review by snnn — APPROVED** [review]

**Review by RyanUnderhill — APPROVED** [review]

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 5]
```diff
@@ -1489,6 +1512,205 @@ struct OrtApi {
...
+   */
+  ORT_API2_STATUS(UseCsrIndices, _Inout_ OrtValue* ort_value, size_t inner_num, _Inout_ int64_t* inner_data, size_t outer_num, _Inout_ int64_t* outer_data);
+
+  /**
+   * The API assigns BlockSparse format indices to the SparseTensor that was created by 
+   * CreateSparseTensorWithValuesAsOrtValue API above. It also sets OrtSparseFormat to 
+   * ORT_SPARSE_BLOCK_SPARSE. The API will not allocate any additional memory for data. The life span of
+   * indices_data buffer must eclipse the lifespan of this OrtValue.
+   * 
+   * \param ort_value - OrtValue instance constructed with CreateSparseTensorWithValuesAsOrtValue
```
[the remark this answers is not in the document]
Comment (yuslepukhin) [reply include/onnxruntime/core/session/onnxruntime_c_api.h:1663]: The correct way is to put it after param keyword

**Review by RyanUnderhill — APPROVED** [review]

**Pull request merged**